accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,245 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+| int? x;           | Ok            | Ok                  | Ok              |
+| final int x;      | Error         | Error               | Ok              |
+| final int? x;     | Error         | Error               | Ok              |
+| late var x;       | Ok            | Ok                  | Ok              |
+| late final x;     | Error         | Ok                  | Ok              |
+| late T x;         | Ok            | Ok                  | Ok              |
+| late final T x;   | Error         | Ok                  | Ok              |
+
+### Local variables and initialization based inference
+
+Local variables with explicitly written types are given the declared types as
+written.  The declared type of the variable is considered a "type of interest"
+in the sense defined in the flow analysis specification.  If the variable has an
+initializer (explicit or implicit) and is not `final`, then the declaration is
+treated as an assignment for the purposes of promotion.
+
+*Treating the declared type of the variable as a "type of interest" implies that
+if the variable is a nullable type, then its non-nullable version is also a type
+of interest.  Treating the initialization as an assignment for the purposes of
+promotion means that initializing a mutable variable declared at type `T?` with
+a value of non-nullable type `T` immediately promotes the variable to the
+non-nullable type.*
+
+```dart
+void test() {
+  int? x = 3; // x is declared at `int?`
+  x.isEven; // Valid, x has been promoted to `int`
+  x = null; // Valid, demotes to the declared type.
+}
+```
+
+Local variables with no explicitly written type but with an initializer are
+given an inferred declared type equal to the type of their initializer.  In the
+case that the type of the initializer is a promoted type variable `X & T`, the
+inferred type of the variable shall be `X`.  However, such a variable shall be
+treated as immediately promoted to `X & T`.
+
+In the null safety release, local variables with no explicitly written type and
+no initializer are subject to further promotion based inference.  The variables
+under consideration here are those declared as `var x;`, `final x;`, `late var
+x;`, or `late final x;`.  We refer to such variables as **untyped variables**.
+
+An **initializing assignment** is an assignment to an **untyped variable** when
+that variable is in the **definitely unassigned** state.  Such an assignment on
+a path shall cause the variable in question to be treated by the flow analysis
+on that path as if it were declared with the assigned expression as its
+initializing expression, up until the next join point in the program.  We refer
+to this inferred type as the **initialization inferred type** of the variable.
+The **initialization inferred type** of the variable is always treated as a
+**type of interest** for the purposes of promotion (including its non-nullable
+version if applicable).
```

Comment:
So
```dart
num n = ...;
late var x;
if (b) { x = 1; x = n; }
```
 is an error (because we can't demote beyond the declared type), but 
```dart
num n = ...; 
late var x; 
if (b) x = 1; 
if (b) x = n;
```
is OK?

The fact that `x` can have many different types in a handful of lines of code, and some assignments will give it a declared type, for a while, and others will give a regular promotion, or a demotion, and so on, makes me think that the notion of initializing assignments may well be counter-productive.

It is very far from "just another syntax for `?:`". I'm tempted to think that when questions come up (like "why does `x` have type `T` here?"), the sane advice will almost always be to eliminate this mechanism to some extent, maybe by initializing `x` immediately, or giving it an explicit type annotation.

Reply:
Maybe.  I don't know how to answer that question without experience with it.  I agree that the example that you give above is a bit weird.  But it's at the intersection of a number of features: a late untyped variable, that is re-assigned when it is in a potentially assigned state.  I'm not willing to throw out all of assignment based promotion because `late` behaves a little oddly in some cases.  In the non-late case, we don't have this anomaly.  We could consider doing something different with late, but so var, this is the best approach that I've found.

Reply:
> I'm tempted to think that when questions come up (like "why does `x` have type `T` here?"), the sane advice will almost always be to eliminate this mechanism to some extent, maybe by initializing `x` immediately, or giving it an explicit type annotation.

Specifically to this, I'm not sure I agree.  The sane advice will be to say "because inference, is it causing a problem?", and as far as I can tell the answer to the latter will almost always be "no".  When do you see the answer being "yes"?

What I have proposed has the nice property that it either tells you it can't infer a type, or it infers a type that I think will usually be what you want.

Reply:
> the sane advice will almost always be to eliminate this mechanism to some extent, maybe by initializing `x` immediately, or giving it an explicit type annotation.

One more comment.  What you're describing is the sane advice *now*.  That is, if you write `var x;` now, you get no static checking at all.  So if your worst case scenario is true (I don't think it is) and out advice becomes "don't use var x;", then we are in no worse of a position from a language standpoint than we are now.

That said, I can see this as an argument for my original proposal, which did not use LUB, and which consequently had the property that each of these variables took on at most one type.  @lrhn @munificent thoughts on this?  I can write up a variant that does not use LUB to see what it would look like concretely.

Reply:
```dart
num n = ...; 
late var x; 
if (b) x = 1; 
if (b) x = n;
```
is actually not OK.
You avoid the error at the join-point after the first `if` because the variable is late, but at the second assignment to `x`, it is no longer definitely unassigned, so it is not a initializing assignment.
The type of `x` at that point is `int`, so the `num` assignment is an error.

Reply:
@lrhn Ah, right, good catch.

Reply:
> so it is not a initializing assignment.

Sure, I constructed the example such that `x = 1` is an initializing assignment and `x = n` is not.

But I got the computation of the pseudo-declared type of `x` wrong: `int` is only valid until the join, but at the join we compute a new pseudo-declared type for `x` (we have the LUB and the "must be the same type" models), and non-assigning branches (which are allowed for a late variable) do not contribute, so it's still `int` when we reach the 2nd `if`.

For instance, if the first `if` is 50 lines of code, and at some point in time it doesn't assign to `x`; then we modify the code and add `x = 1;` somewhere in the first `if`; and then the assignment `x = n` in the second `if` is suddenly treated very differently (because it is no longer an initializing assignment, and in this case it becomes an error).

Do you seriously think that there is any other sane advice for practical Dart development than "go back and write the type of `x` explicitly!"?

----

> That said, I can see this as an argument for my original proposal, which did not use LUB

I think that's at least somewhat more maintainable, and I find it safer to avoid using LUB as far as possible in any case.

Reply:
The main point I was making is that there is a substantial complexity cost that we're imposing on developers, and I think that's been covered in various ways now. So I'll resolve this thread.